### ISSUE TITLE
consul inputs with prometheus outputs

### ISSUE THREAD

**Issue by lastsky** [influxdata/telegraf#2422]
### Relevant telegraf.conf:

~~~
[[outputs.prometheus_client]]
  listen = "127.0.0.1:9126"

[[inputs.consul]]
  ## Most of these values defaults to the one configured on a Consul's agent level.
  ## Optional Consul server address (default: "")
  # address = ""
  ## Optional URI scheme for the Consul server (default: "")
  # scheme = ""
  ## Optional ACL token used in every request (default: "")
  # token = ""
  ## Optional username used for request HTTP Basic Authentication (default: "")
  # username = ""
  ## Optional password used for HTTP Basic Authentication (default: "")
  # password = ""
  ## Optional data centre to query the health checks from (default: "")
  # datacentre = ""
~~~

### System info: versions are latest (or almost latest)

\# apt show telegraf | grep Version
Version: 1.2.1-1

\# telegraf version
Telegraf v1.2.1 (git: release-1.2 3b6ffb344e5c03c1595d862282a6823ecb438cff)

deb url: http://dl.influxdata.com/telegraf/releases/telegraf_1.2.1_amd64.deb

\# cat /etc/*release | grep VERSION=
VERSION="16.04.1 LTS (Xenial Xerus)"

Consul version: 0.7.0

### Steps to reproduce: Usual

1. configure telegraf to input: consul (+ something else) and output: prometheus
2. run telegraf and see alot of useful metrics but NOT consul metrics

### Expected behavior: I have metrics

~~~
$ curl -s 127.0.0.1:9126/metrics | grep consul
# HELP consul_agent_check Is this check passing on this node?
# TYPE consul_agent_check gauge
consul_agent_check{check="serfHealth",node="my_server"} 1
<...>
~~~

### Actual behavior: I do not have consul metrics but have all other metrics

~~~
$ curl -s 127.0.0.1:9126/metrics | grep consul
$
$ curl -s 127.0.0.1:9126/metrics | wc -l
1083
$
$
~~~

### Additional info:
Consul metrics are available and measured:

`telegraf --config /etc/telegraf/telegraf.conf -input-filter consul -test`

~~~
* Plugin: inputs.consul, Collection 1
> consul_health_checks,node=<...> (more than 30 strings)
~~~

### Use case: [Why is this important (helps with prioritizing requests)]

Use case - quick and effectively deploy service monitoring to consul clusters!

**Comment by sparrc** [comment]
that is very strange, are you sure it's not a permissions issue?

can you run telegraf in debug mode and provide output?

```
telegraf --config /etc/telegraf/telegraf.conf -input-filter consul -output-filter prometheus_client --debug
```

**Comment by lastsky** [comment]
Metrics are collected from consul, but not displayed at /metrics url in prometheus output. I see this in large output (skipped above), services are registered in consul, health status is available, telegraf collects them, and collects others at the same moment and test. And telegraf sends to prometheus output all metrics BUT NOT consul metrics :) 

buffer fullness is 10000 metrics but I do not have more than 2000 metrics.

~~~
my_server # telegraf --config /etc/telegraf/telegraf.conf -input-filter consul -output-filter prometheus_client --debug
2017-02-17T00:08:03Z D! Attempting connection to output: prometheus_client
2017-02-17T00:08:03Z D! Successfully connected to output: prometheus_client
2017-02-17T00:08:03Z I! Starting Telegraf (version 1.2.1)
2017-02-17T00:08:03Z I! Loaded outputs: prometheus_client
2017-02-17T00:08:03Z I! Loaded inputs: inputs.consul
2017-02-17T00:08:03Z I! Tags enabled: host=my_server
2017-02-17T00:08:03Z I! Agent Config: Interval:10s, Quiet:false, Hostname:"my_server", Flush Interval:10s
2017-02-17T00:08:20Z D! Output [prometheus_client] buffer fullness: 68 / 10000 metrics.
2017-02-17T00:08:20Z D! Output [prometheus_client] wrote batch of 68 metrics in 649.823µs
2017-02-17T00:08:30Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:08:30Z D! Output [prometheus_client] wrote batch of 34 metrics in 316.539µs
2017-02-17T00:08:40Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:08:40Z D! Output [prometheus_client] wrote batch of 34 metrics in 338.414µs
2017-02-17T00:08:50Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:08:50Z D! Output [prometheus_client] wrote batch of 34 metrics in 346.484µs
2017-02-17T00:09:00Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:09:00Z D! Output [prometheus_client] wrote batch of 34 metrics in 362.114µs
2017-02-17T00:09:10Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:09:10Z D! Output [prometheus_client] wrote batch of 34 metrics in 428.6µs
2017-02-17T00:09:20Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:09:20Z D! Output [prometheus_client] wrote batch of 34 metrics in 331.926µs
2017-02-17T00:09:30Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:09:30Z D! Output [prometheus_client] wrote batch of 34 metrics in 605.326µs
2017-02-17T00:09:40Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:09:40Z D! Output [prometheus_client] wrote batch of 34 metrics in 586.519µs
2017-02-17T00:09:50Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:09:50Z D! Output [prometheus_client] wrote batch of 34 metrics in 599.362µs
2017-02-17T00:10:00Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:10:00Z D! Output [prometheus_client] wrote batch of 34 metrics in 352.012µs
2017-02-17T00:10:10Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:10:10Z D! Output [prometheus_client] wrote batch of 34 metrics in 320.413µs
2017-02-17T00:10:20Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:10:20Z D! Output [prometheus_client] wrote batch of 34 metrics in 536.269µs
2017-02-17T00:10:30Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:10:30Z D! Output [prometheus_client] wrote batch of 34 metrics in 321.598µs
2017-02-17T00:10:40Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:10:40Z D! Output [prometheus_client] wrote batch of 34 metrics in 356.394µs
2017-02-17T00:10:50Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:10:50Z D! Output [prometheus_client] wrote batch of 34 metrics in 316.271µs
2017-02-17T00:11:00Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:11:00Z D! Output [prometheus_client] wrote batch of 34 metrics in 286.898µs
2017-02-17T00:11:10Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:11:10Z D! Output [prometheus_client] wrote batch of 34 metrics in 358.092µs
2017-02-17T00:11:20Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:11:20Z D! Output [prometheus_client] wrote batch of 34 metrics in 331.288µs
2017-02-17T00:11:30Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:11:30Z D! Output [prometheus_client] wrote batch of 34 metrics in 321.105µs
2017-02-17T00:11:40Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:11:40Z D! Output [prometheus_client] wrote batch of 34 metrics in 318.623µs
2017-02-17T00:11:50Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:11:50Z D! Output [prometheus_client] wrote batch of 34 metrics in 404.605µs
2017-02-17T00:12:00Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:12:00Z D! Output [prometheus_client] wrote batch of 34 metrics in 301.703µs
2017-02-17T00:12:10Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:12:10Z D! Output [prometheus_client] wrote batch of 34 metrics in 773.209µs
2017-02-17T00:12:20Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:12:20Z D! Output [prometheus_client] wrote batch of 34 metrics in 339.862µs
2017-02-17T00:12:30Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:12:30Z D! Output [prometheus_client] wrote batch of 34 metrics in 283.586µs
2017-02-17T00:12:40Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:12:40Z D! Output [prometheus_client] wrote batch of 34 metrics in 324.611µs
2017-02-17T00:12:50Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:12:50Z D! Output [prometheus_client] wrote batch of 34 metrics in 301.945µs
2017-02-17T00:13:00Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:13:00Z D! Output [prometheus_client] wrote batch of 34 metrics in 309.272µs
2017-02-17T00:13:10Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:13:10Z D! Output [prometheus_client] wrote batch of 34 metrics in 309.506µs
2017-02-17T00:13:20Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:13:20Z D! Output [prometheus_client] wrote batch of 34 metrics in 331.657µs
2017-02-17T00:13:30Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:13:30Z D! Output [prometheus_client] wrote batch of 34 metrics in 323.769µs
2017-02-17T00:13:40Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:13:40Z D! Output [prometheus_client] wrote batch of 34 metrics in 314.443µs
2017-02-17T00:13:50Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:13:50Z D! Output [prometheus_client] wrote batch of 34 metrics in 293.126µs
2017-02-17T00:14:00Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:14:00Z D! Output [prometheus_client] wrote batch of 34 metrics in 372.506µs
2017-02-17T00:14:10Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:14:10Z D! Output [prometheus_client] wrote batch of 34 metrics in 336.638µs
2017-02-17T00:14:20Z D! Output [prometheus_client] buffer fullness: 34 / 10000 metrics.
2017-02-17T00:14:20Z D! Output [prometheus_client] wrote batch of 34 metrics in 329.418µs
<...>
and so on
~~~

**Comment by lastsky** [comment]
p.s. this behavior is reproduced stabile at 8 different servers in 2 different consul clusters with different OS and telegraf versions

**Comment by lastsky** [comment]
@sparrc added more info about versions.

**Comment by sparrc** [comment]
What happens when you run only with prometheus? Please provide a full config file that can reproduce the issue.

**Comment by lastsky** [comment]
added into gist (alot of output)

https://gist.github.com/lastsky/80926e9a1b1803ee5863a348c20b5328

**Comment by lastsky** [comment]
hi! any news?

**Comment by sparrc** [comment]
can you run as root and see if you still dont have the consul metrics?

```
sudo telegraf --config /etc/telegraf/telegraf.conf -input-filter consul -output-filter prometheus_client --debug
```

**Comment by lastsky** [comment]
still dont have. previously telegraf was started from `root` too.

added into gist (alot of output)

https://gist.github.com/lastsky/c9500c3bd11d221ab6afc30cdbac35b2

**Comment by sparrc** [comment]
Hi @lastsky, the output that you've provided doesn't match your config file. In your prometheus output you have `processes` listed, for example.

I haven't been able to reproduce this issue so I'm going to go ahead and close this.

I would suggest double-checking your environment and setup. Please accept my apologies if you still have issues, but I hope you understand that we can't spend too much time chasing bugs that are not reproducible.

If anyone else has the same issue please comment here and we'll consider re-opening.

**Comment by lastsky** [comment]
Closed??? But `processes` are listed always when at least one `input` + output `[[outputs.prometheus_client]]` are configured.

```process_cpu_seconds_total 33.75
process_max_fds 1024
process_open_fds 7
process_resident_memory_bytes 2.9233152e+07
process_start_time_seconds 1.48730821013e+09
process_virtual_memory_bytes 3.71728384e+08


Surprise!

**Comment by lastsky** [comment]
> I haven't been able to reproduce this issue

Would you like to show me your prometheus output with curl? I will try to find some differences. Maybe dash in service name breaks output or something alike.

**Comment by yangzhares** [comment]
hi guys,

i also hit this issue that consul input-filter can't send metrics to Prometheus output-filter, and i can get metrics that generated by consul input-filter with:
```
telegraf --config=/etc/telegraf/telegraf.conf --config-directory=/etc/telegraf/telegraf.d --input-filter consul --test
* Plugin: inputs.consul, Collection 1
> consul_health_checks,host=mshf1edg01,node=mshf1cnl01,service_name=,check_id=serfHealth check_name="Serf Health Status",service_id="",status="passing" 1491890529000000000
> consul_health_checks,service_name=telegraf,check_id=service:telegraf,host=mshf1edg01,node=mshf1cnl01 status="passing",check_name="Service 'telegraf' check",service_id="telegraf" 1491890529000000000
> consul_health_checks,node=mshf1cnl02,service_name=,check_id=serfHealth,host=mshf1edg01 check_name="Serf Health Status",service_id="",status="passing" 1491890529000000000
> consul_health_checks,service_name=telegraf,check_id=service:telegraf,host=mshf1edg01,node=mshf1cnl02 status="passing",check_name="Service 'telegraf' check",service_id="telegraf" 1491890529000000000
> consul_health_checks,node=mshf1cnl03,service_name=,check_id=serfHealth,host=mshf1edg01 service_id="",status="passing",check_name="Serf Health Status" 1491890529000000000
> consul_health_checks,node=mshf1cnl03,service_name=telegraf,check_id=service:telegraf,host=mshf1edg01 status="passing",check_name="Service 'telegraf' check",service_id="telegraf" 1491890529000000000
> consul_health_checks,node=mshf1edg01,service_name=nomad-client,check_id=289cc7e1737904489a71a4705d50e2dea3a55881,host=mshf1edg01 status="passing",check_name="Nomad Client HTTP Check",service_id="_nomad-client-nomad-client-http" 1491890529000000000
... ...
```
But can't find any metrics about consul input-filter from localhost:9126/metrics:
```
# curl -s localhost:9126/metrics | grep consul_health_checks | wc -l
0
```
As described above, i also can see metrics have been written to Prometheus when enable debug mode:
```
# telegraf --config /etc/telegraf/telegraf.conf -input-filter consul -output-filter prometheus_client --debug
2017-04-11T06:03:33Z D! Attempting connection to output: prometheus_client
2017-04-11T06:03:33Z D! Successfully connected to output: prometheus_client
2017-04-11T06:03:33Z I! Starting Telegraf (version 1.2.1)
2017-04-11T06:03:33Z I! Loaded outputs: prometheus_client
2017-04-11T06:03:33Z I! Loaded inputs: inputs.consul
2017-04-11T06:03:33Z I! Tags enabled: host=mshf1edg01
2017-04-11T06:03:33Z I! Agent Config: Interval:10s, Quiet:false, Hostname:"mshf1edg01", Flush Interval:10s
2017-04-11T06:03:50Z D! Output [prometheus_client] buffer fullness: 116 / 10000 metrics.
2017-04-11T06:03:50Z D! Output [prometheus_client] wrote batch of 116 metrics in 3.233718ms
2017-04-11T06:04:00Z D! Output [prometheus_client] buffer fullness: 58 / 10000 metrics.
2017-04-11T06:04:00Z D! Output [prometheus_client] wrote batch of 58 metrics in 1.764895ms
2017-04-11T06:04:10Z D! Output [prometheus_client] buffer fullness: 58 / 10000 metrics.
2017-04-11T06:04:10Z D! Output [prometheus_client] wrote batch of 58 metrics in 1.719059ms
... ... 
```
Paste telegraf config here for reference:
```
# more telegraf.conf
[global_tags]

# Configuration for telegraf agent
[agent]
  interval = "10s"
  round_interval = true
  metric_batch_size = 1000
  metric_buffer_limit = 10000
  collection_jitter = "0s"
  flush_interval = "10s"
  flush_jitter = "0s"
  precision = ""
  debug = false
  quiet = false
  logfile = ""
  hostname = ""
  omit_hostname = false

###############################################################################
#                                  OUTPUTS                                    #
###############################################################################

[[outputs.prometheus_client]]
  listen = ":9126"
  expiration_interval = "60s"

###############################################################################
#                               CORE INPUTS                                   #
###############################################################################
[[inputs.disk]]
  ignore_fs = ["tmpfs", "devtmpfs"]

[[inputs.diskio]]

[[inputs.kernel]]

[[inputs.mem]]

[[inputs.processes]]

[[inputs.swap]]

[[inputs.system]]

[[inputs.net]]

[[inputs.netstat]]

[[inputs.nstat]]

[[inputs.internal]]

[[inputs.conntrack]]
  files = ["ip_conntrack_count","ip_conntrack_max","nf_conntrack_count","nf_conntrack_max"]
  dirs = ["/proc/sys/net/ipv4/netfilter","/proc/sys/net/netfilter"]

[[inputs.cpu]]
  percpu = true
  totalcpu = true
  collect_cpu_time = false

[[inputs.consul]]
```
Could you help check this issue? thanks in advance.

**Comment by danielnelson** [comment]
I wonder if the consul plugin is creating bad metrics, those tags and fields with empty strings seem suspicious.  I'll try to reproduce.

**Comment by yangzhares** [comment]
it seems consul `serfHealth` has a empty value, maybe can filter out `serfHealth` or mark `serfHealth` with empty value is normal, then check what to happen.

**Comment by OferE** [comment]
Hi,
The same problem happens with output filter graphite as well.
when i run:
telegraf --config=./telegraf.conf  --test 
i get results of the consul health checks, however they are not getting into graphite.

results of the --test run are somthig of the following format:
> consul_health_checks,node=ip-10-0-6-78,service_name=,check_id=serfHealth,class=com.farm,host=ip-10-0-6-78 service_id="",status="passing",check_name="Serf Health Status" 1492947880000000000
> consul_health_checks,check_id=service:alert_ip-10-0-6-78_1492932882411,class=com.farm,host=ip-10-0-6-78,node=ip-10-0-6-78,service_name=alert check_name="Service 'alert' check",service_id="alert_ip-10-0-6-78_1492932882411",status="passing" 1492947880000000000
> consul_health_checks,node=ip-10-0-9-16,service_name=nomad-client,check_id=289cc7e1737904489a71a4705d50e2dea3a55881,class=com.farm,host=ip-10-0-6-78 check_name="Nomad Client HTTP Check",service_id="_nomad-client-nomad-client-http",status="passing" 1492947880000000000
> consul_health_checks,node=ip-10-0-9-16,service_name=nomadui,check_id=57c05859c089de137d3b7f74cee7c482e9c6058a,class=com.farm,host=ip-10-0-6-78 check_name="alive",service_id="_nomad-executor-f3042579-238e-aa39-012f-7e2abb7f4bb7-nomadui-nomadui",status="passing" 1492947880000000000




Thanks,

**Comment by lastsky** [comment]
> I haven't been able to reproduce this issue so I'm going to go ahead and close this.

@sparrc Understood. Actually You hadn't reproducing this at all.

@danielnelson Thanks for reopening!

**Comment by powersj** [comment]
I believe this was fixed with #5880, so I'm going to close this. If this isn't the case please do let us know with logs from the latest Telegraf release.

Thanks!
